unified_planning/grpc/proto_writer.py

```diff
@@ -0,0 +1,341 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import unified_planning.grpc.generated.unified_planning_pb2 as unified_planning_pb2
+from unified_planning.grpc.converter import Converter, handles
+from unified_planning.model.operators import op_to_str
+import unified_planning.model
+import unified_planning.plan
+
+
+def map_operator(op: int) -> str:
+    # TODO: Add support for other types which share the operators space. Eg: REALCONSTANTS, VARIABLE_EXP
+    op = op_to_str(op)
+    if op == "PLUS":
+        return "+"
+    elif op == "MINUS":
+        return "-"
+    elif op == "TIMES":
+        return "*"
+    elif op == "DIV":
+        return "/"
+    elif op == "LE":
+        return "<="
+    elif op == "LT":
+        return "<"
+    elif op == "EQUALS":
+        return "="
+    elif op == "AND":
+        return "and"
+    elif op == "OR":
+        return "or"
+    elif op == "NOT":
+        return "not"
+    elif op == "IMPLIES":
+        return "implies"
+    elif op == "IFF":
+        return "iff"
+    elif op == "EXISTS":
+        return "exists"
+    elif op == "FORALL":
+        return "forall"
+
+    raise ValueError(f"Unknown operator `{op}`")
+
+
+class ProtobufWriter(Converter):
+    @handles(unified_planning.model.Fluent)
+    def _convert_fluent(self, fluent):
+        name = fluent.name()
+        sig = [
+            unified_planning_pb2.Parameter(name="", type=str(t))
+            for t in fluent.signature()
+        ]
+        valType = str(fluent.type())
+        return unified_planning_pb2.Fluent(
+            name=name, value_type=valType, parameters=sig
+        )
+
+    @handles(unified_planning.model.Object)
+    def _convert_object(self, obj):
+        return unified_planning_pb2.ObjectDeclaration(
+            name=obj.name(), type=obj.type().name()
+        )
+
+    @handles(unified_planning.model.FNode)
+    def _convert_fnode(self, exp):
+        payload = exp._content.payload
+        args = exp._content.args
+        atom = None
+        p_type = ""
+        kind = unified_planning_pb2.ExpressionKind.Value("UNKNOWN")
+        arg_list = []
```

Comment:
arg_list might be a confusing name as the expression is either an atom or a list containing the "function" (first element of the list) and its parameters (rest of the list)